File: jclasslib/src/org/gjt/jclasslib/bytecode/ImmediateShortInstruction.java
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.bytecode;

import org.gjt.jclasslib.io.*;
import java.io.*;

/**
    Describes an instruction that is followed by an immediate unsigned short.
 
    @author <a href="mailto:jclasslib@gmx.net">Ingo Kegel</a>
    @version $Revision: 1.2 $ $Date: 2002-02-17 17:32:37 $
*/
public class ImmediateShortInstruction extends AbstractInstruction {

    private int immediateShort;
   
    public int getSize() {
        return super.getSize() + 2;
    }

    public ImmediateShortInstruction(int opcode) {
        super(opcode); 
    }
    
    public ImmediateShortInstruction(int opcode, int immediateShort) {
        super(opcode); 
        this.immediateShort = immediateShort;
    }
    
    /**
        Get the immediate unsigned short of this instruction.
        @return the short
     */
    public int getImmediateShort() {
        return immediateShort;
    }

    /**
        Set the immediate unsigned short of this instruction.
        @param immediateShort the short
     */
    public void setImmediateShort(int immediateShort) {
        this.immediateShort = immediateShort;
    }
    
    public void read(ByteCodeInput in) throws IOException {
        super.read(in);

        immediateShort = in.readUnsignedShort();
    }

    public void write(ByteCodeOutput out) throws IOException {
        super.write(out);

        out.writeShort(immediateShort);
    }
    
}

File: modules/data/src/commonMain/kotlin/bytecode/ImmediateShortInstruction.kt
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license or (at your option) any later version.
*/

package org.gjt.jclasslib.bytecode

import org.gjt.jclasslib.io.CountingDataInput
import org.gjt.jclasslib.io.CountingDataOutput
import org.gjt.jclasslib.structures.ClassFile
import org.gjt.jclasslib.structures.Constant

/**
 * Describes an instruction followed by an immediate unsigned short.
 * @property immediateShort Immediate unsigned short of this instruction.
 */
abstract class ImmediateShortInstruction(opcode: Opcode, var immediateShort: Int) : Instruction(opcode) {

    override val size: Int
        get() = super.size + 2

    override fun read(input: CountingDataInput) {
        super.read(input)

        immediateShort = input.readUnsignedShort()
    }

    override fun write(output: CountingDataOutput) {
        super.write(output)

        output.writeShort(immediateShort)
    }

    override fun isConstantUsed(constant: Constant, classFile: ClassFile): Boolean {
        return opcode != Opcode.SIPUSH && classFile.getConstantPoolIndex(constant) == immediateShort
    }
}

/**
 * Describes an instruction followed by an immediate unsigned short.
 */
class SimpleImmediateShortInstruction(opcode: Opcode, immediateShort: Int = 0) : ImmediateShortInstruction(opcode, immediateShort)
